pkg/controller/common/expectations/per_cluster.go

```diff
@@ -0,0 +1,61 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package expectations
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// ClustersExpectation stores Expectations for several clusters.
+// It is thread-safe, but the underlying per-cluster Expectations is not.
+type ClustersExpectation struct {
+	client   k8s.Client
+	clusters map[types.NamespacedName]*Expectations
+	lock     sync.RWMutex
```

Comment:
I understand we have multiple goroutines updating `clusters`, but a single goroutine accessing data for a particular cluster. If that's correct, should the entire state be divided between clusters (ie. observers, watches and expectations) and access to that be guarded by locks? This would avoid locking in each of those.

Reply:
In general, yes, but on a case by case basis:
* observers are used with concurrent reads and writes, so they need a lock anyway. The observer `Manager` is the global map of Observers for all clusters, which also needs its lock.
* watches are manipulated at the controller-level, not at an individual cluster level in a particular reconciliation. So we need a lock on the cluster map.
* expectations now work the way you describe :)

Reply:
I see, that makes sense. What I had in mind is to remove locking from expectations and move the locking (and access) to a different structure. This would leave expectations agnostic to locks. Since we only could do that to expectations right now, I think it doesn't matter too much, so I'm good.